include/openssl/hpke.h

```diff
@@ -0,0 +1,503 @@
+ * @param pskid is the pskid string for a PSK mode (can be NULL)
+ * @param psklen is the psk length
+ * @param psk is the psk
+ * @param publen is the length of the public (authentication) key
+ * @param pub is the encoded public (authentication) key
+ * @param privlen is the length of the private key
+ * @param priv is the encoded private key
+ * @param evppriv is a pointer to an internal form of private key
+ * @param enclen is the length of the peer's public value
+ * @param enc is the peer's public value
+ * @param cipherlen is the length of the ciphertext
+ * @param cipher is the ciphertext
+ * @param aadlen is the length of the additional data
+ * @param aad is the encoded additional data
+ * @param infolen is the length of the info data (can be zero)
+ * @param info is the encoded info data (can be NULL)
+ * @param seqlen is the length of the sequence data (can be zero)
+ * @param seq is the encoded sequence data (can be NULL)
+ * @param clearlen length of the input buffer for cleartext
+ * @param clear is the encoded cleartext
+ * @return 1 for success, other for error (error returns can be non-zero)
+ */
+int OSSL_HPKE_dec(OSSL_LIB_CTX *libctx,
+                  unsigned int mode, OSSL_HPKE_SUITE suite,
+                  char *pskid, size_t psklen, unsigned char *psk,
+                  size_t publen, unsigned char *pub,
+                  size_t privlen, unsigned char *priv,
+                  EVP_PKEY *evppriv,
+                  size_t enclen, unsigned char *enc,
+                  size_t cipherlen, unsigned char *cipher,
+                  size_t aadlen, unsigned char *aad,
+                  size_t infolen, unsigned char *info,
+                  size_t seqlen, unsigned char *seq,
+                  size_t *clearlen, unsigned char *clear);
+
+/**
+ * @brief generate a key pair
+ *
+ * Used for entities that will later receive HPKE values to
+ * decrypt. Only the KEM from the suite is significant here.
+ * The ``pub` output will typically be published so that
+ * others can encrypt to the private key holder using HPKE.
+ * The ``priv`` output contains the raw private value and
+ * hence is sensitive.
+ *
+ * @param libctx is the context to use (normally NULL)
+ * @param mode is the mode (currently unused)
+ * @param suite is the ciphersuite (currently unused)
+ * @param ikmlen is the length of IKM, if supplied
+ * @param ikm is IKM, if supplied
+ * @param publen is the size of the public key buffer (exact length on output)
+ * @param pub is the public value
+ * @param privlen is the size of the private key buffer (exact length on output)
+ * @param priv is the private key
+ * @return 1 for success, other for error (error returns can be non-zero)
+ */
+int OSSL_HPKE_kg(OSSL_LIB_CTX *libctx,
+                 unsigned int mode, OSSL_HPKE_SUITE suite,
+                 size_t ikmlen, unsigned char *ikm,
+                 size_t *publen, unsigned char *pub,
+                 size_t *privlen, unsigned char *priv);
```

Comment:
It's not strictly necessary. But I figure it should turn out useful for ECH key rotation in haproxy (which I've not yet implemented) because, after initial startup, haproxy doesn't want to depend on reading from disk, so requires use of  some form of IPC (forget if it's socket based or what) thus also calling for passing an ECH private key (which'll be fed into the HPKE API ultimately) as an octet string. As stated though I've not implemented that haproxy key rotation thing yet so it could turn out better to use some other method of going from EVP_PKEY to octets and back to EVP_PLEY, not sure. Advice on other options welcome.